connectors/sources/github.py

```diff
@@ -0,0 +1,252 @@
+
+from connectors.logger import logger
+from connectors.source import BaseDataSource, ConfigurableFieldValueError
+from connectors.utils import CancellableSleeps, RetryStrategy, iso_utc, retryable
+
+RETRIES = 3
+RETRY_INTERVAL = 2
+
+USER_API = "https://api.github.com/user"
+ENDPOINTS = {
+    "ALL_REPOS": "/user/repos?type=all",
+    "REPO": "repos/{repo_name}",
+    "USER": "/user",
+}
+
+
+class GitHubClient:
+    """GitHub Client"""
+
+    def __init__(self, configuration):
+        """Setup the GitHub client.
+
+        Args:
+            configuration (DataSourceConfiguration): Instance of DataSourceConfiguration class.
+        """
+
+        self.session = None
+        self._sleeps = CancellableSleeps()
+        self.configuration = configuration
+        self.repos = self.configuration["repositories"]
+        self.github_token = self.configuration["github_token"]
+        self.headers = {"Authorization": f"Bearer {self.github_token}"}
+
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+    )
+    async def execute_getitem(self, request):
+        """Execute request using getitem method of GitHubAPI.
+
+        Args:
+            request (str): API for the request
+
+        Returns:
+            list: Response of the request
+        """
+        return await self._get_session.getitem(url=request)
+
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+    )
+    async def execute_request(self, method, url):
+        """Execute request using _request method of GitHubAPI.
+
+        Args:
+            request (str): API for the request
+
+        Returns:
+            list: Response of the request
+        """
+        status_code, response_headers, data = await self._get_session._request(
+            method="get", url=USER_API, headers=self.headers
+        )
+        if status_code == 200:
+            return response_headers
+        data = data.decode("utf-8")
+        message = json.loads(data)
+        raise Exception(
+            f"Something went wrong while executing request Exception: {message['message']}"
+        )
+
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+    )
+    async def execute_getiter(self, request):
+        """Execute request using getiter method of GitHubAPI.
+
+        Args:
+            request (str): API for the request
+
+        Yields:
+            dict: Response of the request
+        """
+        async for response in self._get_session.getiter(url=request):
+            yield response
+
+    async def filter_repos(self):
+        try:
+            logger.debug("Filtering repositories")
+            public_repos, configured_repos, users_repos = [], [], []
+            for repo in self.repos:
+                if "/" in repo:
```

Comment:
Is a repository always public, if it contains a `/`?

Reply:
If the user wants to fetch repos of a different user, then the user needs to configure the repo name as `owner/repo_name`. As `public_repos` is misleading, I will change the variable name to `external_repos`.

Reply:
sounds good